internal/coss/storage/environments/git/azure/azure.go

```diff
@@ -0,0 +1,252 @@
+	}, nil
+}
+
+type SCM struct {
+	logger      *zap.Logger
+	client      Client
+	repoOwner   string
+	repoProject string
+	repoName    string
+	baseURL     string
+}
+
+func (s *SCM) ListChanges(ctx context.Context, req git.ListChangesRequest) (*environments.ListBranchedEnvironmentChangesResponse, error) {
+	changes := []*environments.Change{}
+	includeLinks := true
+	commits, err := s.client.GetCommits(ctx, azuregit.GetCommitsArgs{
+		RepositoryId: &s.repoName,
+		Project:      &s.repoProject,
+		SearchCriteria: &azuregit.GitQueryCommitsCriteria{
+			IncludeLinks: &includeLinks,
+			ItemVersion: &azuregit.GitVersionDescriptor{
+				Version:     &req.Base,
+				VersionType: &azuregit.GitVersionTypeValues.Branch,
+			},
+			CompareVersion: &azuregit.GitVersionDescriptor{
+				Version:     &req.Head,
+				VersionType: &azuregit.GitVersionTypeValues.Branch,
+			},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to compare branches: %w", err)
+	}
+
+	for _, commit := range *commits {
+		change := &environments.Change{
+			Revision: *commit.CommitId,
+			Message:  *commit.Comment,
+			ScmUrl:   commit.RemoteUrl,
+		}
+		if commit.Author != nil {
+			if commit.Author.Date != nil {
+				change.Timestamp = commit.Author.Date.Time.Format(time.RFC3339)
+			}
+			change.AuthorName = commit.Author.Name
+			change.AuthorEmail = commit.Author.Email
+		}
+		changes = append(changes, change)
+	}
+	return &environments.ListBranchedEnvironmentChangesResponse{
+		Changes: changes,
+	}, nil
+}
+
+func (s *SCM) Propose(ctx context.Context, req git.ProposalRequest) (*environments.EnvironmentProposalDetails, error) {
+	s.logger.Info("proposing pull request", zap.String("base", req.Base), zap.String("head", req.Head), zap.String("title", req.Title), zap.Bool("draft", req.Draft))
+	sourceRefName := fmt.Sprintf("refs/heads/%s", req.Head)
+	targetRefName := fmt.Sprintf("refs/heads/%s", req.Base)
+	pr, err := s.client.CreatePullRequest(ctx, azuregit.CreatePullRequestArgs{
+		GitPullRequestToCreate: &azuregit.GitPullRequest{
+			Title:         &req.Title,
+			Description:   &req.Body,
+			IsDraft:       &req.Draft,
+			SourceRefName: &sourceRefName,
+			TargetRefName: &targetRefName,
+		},
+		RepositoryId: &s.repoName,
+		Project:      &s.repoProject,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pull request: %w", err)
+	}
+	s.logger.Info("pull request created", zap.String("pr", *pr.Url), zap.String("state", string(*pr.Status)))
+	return &environments.EnvironmentProposalDetails{
+		Url:   *pr.Url,
+		State: environments.ProposalState_PROPOSAL_STATE_OPEN,
+	}, nil
+}
+
+func (s *SCM) ListProposals(ctx context.Context, env serverenvs.Environment) (map[string]*environments.EnvironmentProposalDetails, error) {
+	var (
+		details = map[string]*environments.EnvironmentProposalDetails{}
+		prs     = &prs{ctx: ctx, client: s.client, repoProject: s.repoProject, repoName: s.repoName, base: env.Configuration().Ref}
```

Comment:
To make it work, I have to use ` env.Configuration().Ref` instead of `env.Key()`. I don't know if there is something different or there is a bug somewhere.

Reply:
Gitea probably is buggy.